pennylane/devices/device_api.py

```diff
@@ -0,0 +1,447 @@
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        # each instance should have its own Tracker.
+        self.tracker = Tracker()
+
+    def preprocess(
+        self, qscript: QScript_or_QScriptBatch, execution_config=None
+    ) -> Tuple[QScriptBatch, Callable]:
+        """Device preprocessing function.
+
+        Args:
+            qscript (Union[QuantumScript, Sequence[QuantumScript]]): The circuit or a batch of circuits to preprocess
+                before execution on the device
+            execution_config (ExecutionConfig): A datastructure describing the parameters needed to fully describe
+                the execution. Includes such information as shots.
+
+        Returns:
+            Sequence[QuantumScript], Callable: QuantumScripts that the device can natively execute without
+            error and a postprocessing function to be called after execution.
+
+        Raises:
+            Exception: An exception is raised if the input cannot be converted into a form supported by the device.
+
+        This function is tracked by machine learning interfaces and should be fully differentiable.
+
+        Preprocessing steps may include:
+        - expansion to :class:`~.Operator`'s and :class:`~.MeasurementProcess` objects supported by the device.
+        - splitting a quantum script with the measurement of non-commuting observables or Hamiltonians into multiple executions
+        - splitting quantum scripts with batched parameters into multiple executions
+
+        This step may also validate parameters of the ``execution_config``. For example, it could raise an error if the
+        device doesn't support finite shots but the ``execution_config`` requests it.
+
+        Processing steps can depend on the requested gradient method.
+
+        """
+
+        def blank_postprocessing_fn(res):
+            """Identity postprocessing function created in QuantumDevice preprocessing.
+
+            Args:
+                res (tensor-like): A result object
+
+            Returns:
+                tensor-like: The function input.
+
+            """
+            return res
+
+        qscript_batch = [qscript] if isinstance(qscript, QuantumScript) else qscript
+        return qscript_batch, blank_postprocessing_fn
+
+    @abc.abstractmethod
+    def execute(self, qscripts: QScript_or_QScriptBatch, execution_config=None):
+        """Execute a Quantum Script and turn it into a result.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            A numeric result of the computation.
+
+        **Execution Config:**
+
+        The execution config can be:
+
+        * ``None``: device defaults should be used
+
+        * A single ``ExecutionConfig`` datastructure: All scripts in the batch use the same configuration
```

Comment:
Maybe out of scope: could the execution config be attached to the tape?

Reply:
See [ADR PR#53](https://github.com/PennyLaneAI/adrs/pull/53/files) for more discussion on this. The tape is "what is to be executed" and the execution config is "how it is to be executed".  While we could eventually have a class that combines these two things, we would still want two separate classes for the what and the how.